mypy/typeanal.py

```diff
@@ -1354,7 +1354,49 @@ def visit_union_type(self, t: UnionType) -> Type:
             and not self.options.python_version >= (3, 10)
         ):
             self.fail("X | Y syntax for unions requires Python 3.10", t, code=codes.SYNTAX)
-        return UnionType(self.anal_array(t.items), t.line, uses_pep604_syntax=t.uses_pep604_syntax)
+        items = self.anal_array(t.items)
+        # Check for invalid quoted type inside union syntax.
+        if (
+            t.original_str_expr is None
+            and not self.api.is_stub_file
+            and (not self.api.is_future_flag_set("annotations") or self.defining_alias)
+        ):
```

Comment:
Style nit: This new if statement now accounts for most the code in this function. What about moving it to a helper function to make the basic logic clearer?